docs/proposals/stateful-apps.md

```diff
@@ -0,0 +1,406 @@
+components. That does not mean that all stateful applications *must* use StatefulSets, but the tradeoffs
+in this resource are intended to facilitate holding state in the cluster.
+
+
+## Use Cases
+
+The software listed below forms the primary use-cases for a StatefulSet on the cluster - problems encountered
+while adapting these for Kubernetes should be addressed in a final design.
+
+* Quorum with Leader Election
+  * MongoDB - in replica set mode forms a quorum with an elected leader, but instances must be preconfigured
+    and have stable network identities.
+  * ZooKeeper - forms a quorum with an elected leader, but is sensitive to cluster membership changes and
+    replacement instances *must* present consistent identities
+  * etcd - forms a quorum with an elected leader, can alter cluster membership in a consistent way, and
+    requires stable network identities
+* Decentralized Quorum
+  * Cassandra - allows flexible consistency and distributes data via innate hash ring sharding, is also
+    flexible to scaling, more likely to support members that come and go. Scale down may trigger massive
+    rebalances.
+* Active-active
+  * Galera - has multiple active masters which must remain in sync
+* ???
+
+
+## Background
+
+Replica sets are designed with a weak guarantee - that there should be N replicas of a particular
+pod template. Each pod instance varies only by name, and the replication controller errs on the side of
+ensuring that N replicas exist as quickly as possible (by creating new pods as soon as old ones begin graceful
+deletion, for instance, or by being able to pick arbitrary pods to scale down). In addition, pods by design
+have no stable network identity other than their assigned pod IP, which can change over the lifetime of a pod
+resource. ReplicaSets are best leveraged for stateless, shared-nothing, zero-coordination,
+embarassingly-parallel, or fungible software.
+
+While it is possible to emulate the guarantees described above by leveraging multiple replication controllers
+(for distinct pod templates and pod identities) and multiple services (for stable network identity), the
+resulting objects are hard to maintain and must be copied manually in order to scale a cluster.
+
+By constrast, a DaemonSet *can* offer some of the guarantees above, by leveraging Nodes as stable, long-lived
+entities. An administrator might choose a set of nodes, label them a particular way, and create a
+DaemonSet that maps pods to each node. The storage of the node itself (which could be network attached
+storage, or a local SAN) is the persistent storage. The network identity of the node is the stable
+identity. However, while there are examples of clustered software that benefit from close association to
+a node, this creates an undue burden on administrators to design their cluster to satisfy these
+constraints, when a goal of Kubernetes is to decouple system administration from application management.
+
+
+## Design Assumptions
+
```

Comment:
@bprashanth Out of curiosity - does the "limited scaling" also includes limited down-scaling? Is it possible to say I can't go lower than 3-4 replicas?